Erro There is not enough space on the disk 232b1dca347551c734a5f1e66957bf4a64cf8fc2 v2.7.4

br.univali.ps.nucleo.ExcecaoAplicacao: java.io.IOException: There is not enough space on the disk
	at br.univali.ps.nucleo.TratadorExcecoes.transformarExcecao(TratadorExcecoes.java:267)
	at br.univali.ps.nucleo.TratadorExcecoes.exibirExcecao(TratadorExcecoes.java:42)
	at br.univali.ps.ui.abas.AbaCodigoFonte.salvaArquivoRecuperavel(AbaCodigoFonte.java:965)
	at br.univali.ps.ui.abas.AbaCodigoFonte.access$3200(AbaCodigoFonte.java:105)
	at br.univali.ps.ui.abas.AbaCodigoFonte$25.propertyChange(AbaCodigoFonte.java:1223)
	at java.beans.PropertyChangeSupport.fire(PropertyChangeSupport.java:335)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:328)
	at java.beans.PropertyChangeSupport.firePropertyChange(PropertyChangeSupport.java:263)
	at br.univali.ps.ui.rstautil.PortugolParser.parse(PortugolParser.java:196)
	at org.fife.ui.rsyntaxtextarea.ParserManager.actionPerformed(ParserManager.java:157)
	at javax.swing.Timer.fireActionPerformed(Timer.java:313)
	at javax.swing.Timer$DoPostEvent.run(Timer.java:245)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.io.IOException: There is not enough space on the disk
	at java.io.FileOutputStream.writeBytes(Native Method)
	at java.io.FileOutputStream.write(FileOutputStream.java:326)
	at sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:221)
	at sun.nio.cs.StreamEncoder.implFlushBuffer(StreamEncoder.java:291)
	at sun.nio.cs.StreamEncoder.implFlush(StreamEncoder.java:295)
	at sun.nio.cs.StreamEncoder.flush(StreamEncoder.java:141)
	at java.io.OutputStreamWriter.flush(OutputStreamWriter.java:229)
	at java.io.BufferedWriter.flush(BufferedWriter.java:254)
	at br.univali.ps.ui.utils.FileHandle.save(FileHandle.java:24)
	at br.univali.ps.ui.utils.FileHandle.save(FileHandle.java:15)
	at br.univali.ps.ui.abas.AbaCodigoFonte.salvaArquivoRecuperavel(AbaCodigoFonte.java:961)
	... 23 more

--- a/ide/src/main/java/br/univali/ps/nucleo/TratadorExcecoes.java
+++ b/ide/src/main/java/br/univali/ps/nucleo/TratadorExcecoes.java
@@ -142,8 +142,8 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
     @Override
     public void uncaughtException(Thread thread, Throwable excecao)
     {
-        final ExcecaoAplicacao excecaoAplicacao = new ExcecaoAplicacao(excecao, ExcecaoAplicacao.Tipo.ERRO_PROGRAMA);        
-        
+        final ExcecaoAplicacao excecaoAplicacao = new ExcecaoAplicacao(excecao, ExcecaoAplicacao.Tipo.ERRO_PROGRAMA);
+            
         if ((excecao instanceof ClassNotFoundException) || (excecao instanceof NoClassDefFoundError))
         {
             String mensagem;
@@ -165,15 +165,46 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
         {
             LOGGER.log(Level.WARNING, "Exceção não identificada", excecao);
         }
-        else if (silenciador_de_excecoes())
-        {
-            //Shhhh, você não viu nada aqui, continue andando...
-        }
+        else if (silenciador_de_excecoes()){}
         else
         {
             exibirExcecao(excecaoAplicacao);
         }
     }
+    
+    private ExcecaoAplicacao excecaoExterna(Throwable excecao)
+    {
+        String mensagem = excecao.getLocalizedMessage()==null?"":excecao.getLocalizedMessage();
+        String erro = "";
+        
+        String[] search = new String[1];
+        search[0] = "There is not enough space on the disk";
+        if(silenciador_de_excecoes_and(mensagem, search))
+        {
+            erro = "Não há espaço no disco!";
+        }
+        
+        search = new String[1];
+        search[0] = "sendo usado por outro processo";
+        if(silenciador_de_excecoes_and(mensagem, search))
+        {
+            erro = "O arquivo está sendo utilizado em outro processo. Não é possível altera-lo!";
+        }
+        
+        search = new String[1];
+        search[0] = "O provedor do arquivo de nuvem";
+        if(silenciador_de_excecoes_and(mensagem, search))
+        {
+            erro = "O seu provedor de arquivos OneDrive está impedindo o Portugol de fazer modificações. Por favor reinicie ele.";
+        }
+        
+        if(erro.isEmpty())
+        {
+            return null;
+        }
+        return new ExcecaoAplicacao(erro, excecao, ExcecaoAplicacao.Tipo.ERRO_USUARIO);
+    }
+    
     private boolean silenciador_de_excecoes(){
         String mensagem = fluxoSaida.getFullStack();
         //ignorar até JDK ser atualizada para versão 9
@@ -184,9 +215,8 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
             return true;
         
         //https://github.com/UNIVALI-LITE/Portugol-Studio/issues/486
-        search = new String[2];
-        search[0] = "java.lang.NullPointerException: peer";
-        search[1] = "Caused by: java.lang.NullPointerException: peer";
+        search = new String[1];
+        search[0] = "peer";
         if(silenciador_de_excecoes_and(mensagem, search))
             return true;
         
@@ -199,8 +229,8 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
         
         //issue #495
         search = new String[2];
-        search[0] = "org.fife.ui.rsyntaxtextarea.ParserManager.noticeContainsPosition(ParserManager.java:531)";
-        search[1] = "javax.swing.ToolTipManager$insideTimerAction.actionPerformed(ToolTipManager.java:663)";
+        search[0] = "org.fife.ui.rsyntaxtextarea.ParserManager.noticeContainsPosition";
+        search[1] = "javax.swing.ToolTipManager$insideTimerAction.actionPerformed";
         if(silenciador_de_excecoes_and(mensagem, search))
             return true;
         return false;
@@ -276,11 +306,15 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
 
     private ExcecaoAplicacao transformarExcecao(Exception excecao)
     {
+        ExcecaoAplicacao ea = excecaoExterna(excecao);
+        if(ea!=null){
+            return ea;
+        }
         if (!(excecao instanceof ExcecaoAplicacao))
         {
             excecao = new ExcecaoAplicacao(excecao, ExcecaoAplicacao.Tipo.ERRO_PROGRAMA);
         }
-
+        
         return (ExcecaoAplicacao) excecao;
     }
 }

--- a/ide/src/main/java/br/univali/ps/nucleo/TratadorExcecoes.java
+++ b/ide/src/main/java/br/univali/ps/nucleo/TratadorExcecoes.java
@@ -172,39 +172,6 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
         }
     }
     
-    private ExcecaoAplicacao excecaoExterna(Throwable excecao)
-    {
-        String mensagem = excecao.getLocalizedMessage()==null?"":excecao.getLocalizedMessage();
-        String erro = "";
-        
-        String[] search = new String[1];
-        search[0] = "There is not enough space on the disk";
-        if(silenciador_de_excecoes_and(mensagem, search))
-        {
-            erro = "Não há espaço no disco!";
-        }
-        
-        search = new String[1];
-        search[0] = "sendo usado por outro processo";
-        if(silenciador_de_excecoes_and(mensagem, search))
-        {
-            erro = "O arquivo está sendo utilizado em outro processo. Não é possível altera-lo!";
-        }
-        
-        search = new String[1];
-        search[0] = "O provedor do arquivo de nuvem";
-        if(silenciador_de_excecoes_and(mensagem, search))
-        {
-            erro = "O seu provedor de arquivos OneDrive está impedindo o Portugol de fazer modificações. Por favor reinicie ele.";
-        }
-        
-        if(erro.isEmpty())
-        {
-            return null;
-        }
-        return new ExcecaoAplicacao(erro, excecao, ExcecaoAplicacao.Tipo.ERRO_USUARIO);
-    }
-    
     private boolean silenciador_de_excecoes(){
         String mensagem = fluxoSaida.getFullStack();
         //ignorar até JDK ser atualizada para versão 9
@@ -306,10 +273,6 @@ public final class TratadorExcecoes implements Thread.UncaughtExceptionHandler
 
     private ExcecaoAplicacao transformarExcecao(Exception excecao)
     {
-        ExcecaoAplicacao ea = excecaoExterna(excecao);
-        if(ea!=null){
-            return ea;
-        }
         if (!(excecao instanceof ExcecaoAplicacao))
         {
             excecao = new ExcecaoAplicacao(excecao, ExcecaoAplicacao.Tipo.ERRO_PROGRAMA);

--- a/ide/src/main/java/br/univali/ps/ui/abas/AbaCodigoFonte.java
+++ b/ide/src/main/java/br/univali/ps/ui/abas/AbaCodigoFonte.java
@@ -955,10 +955,8 @@ public final class AbaCodigoFonte extends Aba implements PortugolDocumentoListen
                 String texto = documento.getText(0, documento.getLength());
                 texto = inserirInformacoesPortugolStudio(texto);
                 FileHandle.save(texto, getArquivoComExtensaoRecuperacao(arquivoRecuperavel));
-            } catch (BadLocationException ex) {
-                PortugolStudio.getInstancia().getTratadorExcecoes().exibirExcecao(ex);
             } catch (Exception ex) {
-                PortugolStudio.getInstancia().getTratadorExcecoes().exibirExcecao(ex);
+                LOGGER.log(Level.WARNING, null, ex);
             }
         }
     }

--- a/ide/src/main/java/br/univali/ps/ui/utils/FileHandle.java
+++ b/ide/src/main/java/br/univali/ps/ui/utils/FileHandle.java
@@ -34,6 +34,13 @@ public final class FileHandle
                     if(!pasta.contains("recuperavel"))
                     throw new ExcecaoAplicacao("Não há espaço no disco", ExcecaoAplicacao.Tipo.ERRO_USUARIO);
                 }
+                else if (e.getMessage().contains("used by another process") || e.getMessage().contains("sendo usado por outro processo")) {
+                    String arquivo = file.getName();
+                    throw new ExcecaoAplicacao("O arquivo "+arquivo+" está sendo utilizado em outro processo. Não é possível altera-lo!", ExcecaoAplicacao.Tipo.ERRO_USUARIO);
+                }
+                else if (e.getMessage().contains("cloud file provider") || e.getMessage().contains("O provedor do arquivo de nuvem")) {
+                    throw new ExcecaoAplicacao("O seu provedor de arquivos OneDrive está impedindo o Portugol de fazer modificações. Por favor reinicie ele.", ExcecaoAplicacao.Tipo.ERRO_USUARIO);
+                }
                 else {
                     throw e;
                 }

